include/libcyphal/transport/udp/udp_transport_impl.hpp

```diff
@@ -0,0 +1,708 @@
+        media_array.reserve(media_count);
+        if (media_array.capacity() >= media_count)
+        {
+            std::size_t index = 0;
+            for (IMedia* const media_interface : media_interfaces)
+            {
+                if (media_interface != nullptr)
+                {
+                    IMedia& media = *media_interface;
+                    media_array.emplace_back(index, media, local_node_id_, tx_capacity, udp_mem_res);
+                    index++;
+                }
+            }
+            CETL_DEBUG_ASSERT(index == media_count, "");
+            CETL_DEBUG_ASSERT(media_array.size() == media_count, "");
+        }
+
+        return media_array;
+    }
+
+    /// @brief Tries to run an action with media and its TX socket (the latter one is made on demand if necessary).
+    ///
+    template <typename Action>
+    CETL_NODISCARD cetl::optional<AnyError> withEnsureMediaTxSocket(Media& media, Action&& action)
+    {
+        using ErrorReport = TransientErrorReport::MediaMakeTxSocket;
+
+        if (!media.tx_socket_ptr())
+        {
+            auto maybe_tx_socket = media.interface().makeTxSocket();
+            if (auto* const error = cetl::get_if<cetl::variant<MemoryError, PlatformError>>(&maybe_tx_socket))
+            {
+                return tryHandleTransientMediaError<ErrorReport>(media, std::move(*error), media.interface());
+            }
+
+            media.tx_socket_ptr() = cetl::get<UniquePtr<ITxSocket>>(std::move(maybe_tx_socket));
+            if (!media.tx_socket_ptr())
+            {
+                return tryHandleTransientMediaError<ErrorReport, cetl::variant<MemoryError>>(media,
+                                                                                             MemoryError{},
+                                                                                             media.interface());
+            }
+        }
+
+        return std::forward<Action>(action)(media, *media.tx_socket_ptr());
+    }
+
+    CETL_NODISCARD cetl::optional<AnyError> ensureMediaTxSockets()
+    {
+        for (Media& media : media_array_)
+        {
+            cetl::optional<AnyError> any_error =
+                withEnsureMediaTxSocket(media, [](auto&, auto&) -> cetl::nullopt_t { return cetl::nullopt; });
+            if (any_error.has_value())
+            {
+                return any_error;
+            }
+        }
+
+        return cetl::nullopt;
+    }
+
+    void flushUdpardTxQueue(UdpardTx& udpard_tx) const
+    {
+        while (const UdpardTxItem* const maybe_item = ::udpardTxPeek(&udpard_tx))
+        {
+            UdpardTxItem* const item = ::udpardTxPop(&udpard_tx, maybe_item);
+            ::udpardTxFree(memoryResources().fragment, item);
+        }
+    }
+
+    /// @brief Runs transmission loop for each redundant media interface.
+    ///
+    CETL_NODISCARD cetl::optional<AnyError> runMediaTransmit(const TimePoint now)
```

Comment:
Oh dear. We've missed something important in `IRunnable`. We don't have a timeout. This means we cannot schedule this work on single-threaded systems nor can we implement variable schedulers in the future per https://forum.opencyphal.org/t/design-review-execution-model-for-libcyphal/1918 . Even if we do add a timeout we haven't defined how to get a clock for this implementation yet, or have we?

Reply:
Scott, is it (the timeout) is really needed? I thought that we have requirement that no entity inside of a `run` is allowed to block, so if you run a runnable frequently enough there will be not too much to do on each such run. With this approach I don't see why it can't be run on a single-threaded systems (aka super-loop).

Also, I believe that introducing such timeout (for a single run, if I guessed correctly what you meant) will mean that you need to bring/pass it almost everywhere inside (essentially to any method/entity which might do any loop OR anything at all which might overrun allowed slot of time). I believe it will pollute codebase a lot, and will make it harder unit test (b/c you need to cover any potential break condition due to timeout). Error handling probably will be affected as well...

Here is another idea - what if (user-provided) `IMedia` implementation could (if needed) also monitor current allowed time-slot for a run, and if it "sees" (on its `push`/`send`/`pop` etc invocations) that the time-slot is or close enough to be exhausted than it will start quickly return "not ready" - this will quickly unwind/break whatever loops we have in libcyphal transport implementation. Also such user-provided `IMedia` implementation probably could have sense of its timing, so below concern won't be libcyphals one:
>_we haven't defined how to get a clock for this implementation yet_

Reply:
Anyway, I believe Scott this discussion is beyond this PR.

Reply:
@thirtytwobits please check this and the following section titled "Second proposal: Stateful": https://github.com/OpenCyphal-Garage/libcyphal/blob/main/docs/design/readme.md#execution-flow-and-the-runner-interface

>This approach is not compatible with the one-thread-per-runnable execution strategy because all entities that share a given `IMultiplexer` will necessarily have to share the same executor. It will still be possible to run one transport instance per thread, though.

A transport (or some presentation-layer entity) can run in a separate thread similarly to what you described in that thread; where this solution differs is that we never block inside LibCyphal (it's a hard policy mentioned in the introductory part of the document -- everything is nonblocking) but instead we block in the platform-specific code implementing `IMultiplexer`. The client code will block on its multiplexer implementation until there is an event to handle or the next polling deadline is reached, then `IRunnable::run` will be invoked once to handle the events, then the execution thread is blocked again until the next event.

A POSIX-specific implementation of `IMultiplexer` will be probably based on `poll`; such an implementation can provide a utility method that invokes `run` in a loop continulusly, timed by the `poll` call.

Reply:
We can move this discussion to another forum but my concern is that the bounds of these while-loops is affected by the amount of I/O which couples the WCET of each IRunnable to receive rates (coupling between vehicle components) and publish rates (coupling between software components). How do we ensure that these loops have a bounded execution time without using a time bound?

Reply:
> Also, I believe that introducing such timeout (for a single run, if I guessed correctly what you meant) will mean that you need to bring/pass it almost everywhere inside (essentially to any method/entity which might do any loop OR anything at all which might overrun allowed slot of time).

Not necessarily. What we need to ensure is that any loop which can change in size based on external inputs can be bounded, especially if the bounds can change while we are in the loop. In this example, `::udpardTxPeek(&media.udpard_tx())` is bounding the loop which – theoretically if not practically – may continue to grow indefinitely unless we're enforcing single-thread access to each udpard instance?

Reply:
> Here is another idea - what if (user-provided) IMedia implementation could (if needed) also monitor current allowed time-slot for a run, and if it "sees" (on its push/send/pop etc invocations) that the time-slot is or close enough to be exhausted then it will start quickly return "not ready" or "nullopt" data 

Hmm. How would the IMedia know that a new run-cycle had started?

Reply:
>We can move this discussion to another forum but my concern is that the bounds of these while-loops is affected by the amount of I/O which couples the WCET of each IRunnable to receive rates (coupling between vehicle components) and publish rates (coupling between software components). How do we ensure that these loops have a bounded execution time without using a time bound?

Ah, I understand now what we're after here. Per the design proposed in the design doc, once the multiplexer has signaled an event, we will keep `run`ning the `IRunnable`s related to that multiplexer until all work is done. Indeed, there is no way to communicate a deadline --- the call will not return until the RX queue in the media is emptied and there are no pending TX frames left OR the media TX queue is full.

We could retrofit a deadline here quite easily, but think about the temporal granularity here. Our runnables do the job in small chunks --- we need to either push or pop a frame per iteration. Is it acceptable to honor the deadline with the granularity that is defined by the time it takes to process one work item (one frame, etc)?

At the transport layer we have no activities that could take an undefined amount of time to complete (the time complexity of each operation is well-defined in the lizards), but then on the presentation layer we may be dealing with more complex entities such as user callbacks and so on. Is it still acceptable to provide the deadline guarantees whose resolution is essentially dependent on the user code?

Reply:
>  Is it acceptable to honor the deadline with the granularity that is defined by the time it takes to process one work item (one frame, etc)?

Yes. Absolutely. This is the dominant term so as long as we bound the WCET to ±1 unit of work we'll meet our objectives of bounded-realtime if not hard-realtime.

Reply:
Moved into #364